### ISSUE TITLE
Check if its possible to reduce the docker image size of pact broker ?

### ISSUE THREAD

**Issue by shashidesai** [DiUS/pact_broker-docker#52]
* Current docker image size of **dius/pact-broker** = **819MB**

The postgres docker image used with pact broker is just 270MB. So relative to that pact broker seems to be on the heavier side. 
Was wondering if its possible to reduce the docker image size for pact broker ?

**Comment by bethesque** [comment]
That does seem ridiculous! There's not that much going on in it. I wonder if it's all the extra rubies that come installed with phusion-passenger.

**Comment by bethesque** [comment]
I'm assuming the 819MB you quote above is the uncompressed size.

The compressed size of the ruby passenger base image is [244MB](https://hub.docker.com/r/phusion/passenger-ruby24/tags/). The compressed size of the pact-broker image is [294MB](https://hub.docker.com/r/dius/pact-broker/tags/). So the bulk of it comes from the base image. I'm not a docker expert, so I'm happy to take a PR if you could do some research into what might be removed from the passenger base image.

**Comment by bethesque** [comment]
This might contain some useful info: https://semaphoreci.com/blog/2016/12/13/lightweight-docker-images-in-5-steps.html

**Comment by shashidesai** [comment]
@bethesque - I am not aware of the ruby world, but I could give it a shot. if you have any links which I could use, please do let me know. 
I can take this up if its fine with you guys ?

**Comment by bethesque** [comment]
That would be awesome, thanks! I

**Comment by DerKnecht** [comment]
I'm sure you guys, especially shashidesai, already know this, but in most cases you should prefer the COPY command over ADD.
https://www.ctl.io/developers/blog/post/dockerfile-add-vs-copy/

**Comment by bethesque** [comment]
I'm a total docker rookie. Very happy to accept PRs from people who know more than I do.

**Comment by k-ong** [comment]
the pact-broker is built from the phusion passenger container:
phusion/passenger-ruby24:0.9.26 is at 659MB uncompressed.

It definitely seems to be on the heavier side, but comparing it to the postgres image is slightly misleading.

**Comment by shashidesai** [comment]
Yep, I get that comparing it to postgres is slightly misleading but for a small app like pact broker I thought 819MB was too large an image.
I was trying to experiment by not using the phusion/passenger image (which is already heavier), let me know what you think about that ? 
If it doesn't make sense, then we could skip and close this issue ?

**Comment by DerKnecht** [comment]
This indeed makes sense. I would like to have a lightweight docker image, too.
The phusion is based on ubuntu, which is kinda heavy.
I would suggest something which is somewhat time-consuming but a one time job:
Based on Alpine or Debian creating an minimalistic image with only the dependencies that are needed for pact broker. (Alpine is about 5MB or so...) I'm not a docker pro, though I have read lots about this way of creating an image for own purposes and I think this is a good way to get rid of the baggage we got here from the base image.

**Comment by shashidesai** [comment]
@DerKnecht - thanks for the insightful info.

**Comment by bethesque** [comment]
I'm all for making the image as small as possible, but I'm curious as to what the impact is of the larger one, given that it only gets updated every few weeks or so at the moment. Can someone explain the use case? Just to let you know, this is not something I have time to work on at the moment (every bit of spare time I have goes to PB features), but I'm very happy to accept PRs.

**Comment by shashidesai** [comment]
It's not much of an issue, it was just an observation. Anyway it's not that we download the image every now and then. Once it's setup and deployed, we don't need to update the image unless there is an update.
I understand about your time, as this is not a priority right now, that's the reason I mentioned I can pitch in for it as it's not something urgent and can be done at leisure time.

**Comment by mefellows** [comment]
FWIW you could take a look at https://github.com/pact-foundation/pact-mock-service-docker/blob/master/Dockerfile and https://github.com/DiUS/pact-provider-verifier-docker/blob/master/Dockerfile for inspiration (we converted them to Alpine and delt with the Ruby issues). 

The broker image is a little more complicated with Nginx etc., so if time permits we can look into it but as Beth said, it's not a big issue/priority for us.

PR's absolutely welcome though :)

**Comment by k-ong** [comment]
I agree it does have some complexities around Nginx as we probably need to compile from source.
and with that said, I might be doing something wrong but my image with alpine is growing to ~600Mb :(

I will be grateful to see if others have done a better job :)

**Comment by k-ong** [comment]
wip broker docker image with alpine: https://github.com/DiUS/pact_broker-docker/tree/broker-alpine

the image however, is sitting at 742Mb. will now attempt to slim it down but also looking for feedback/comments

**Comment by mefellows** [comment]
Having spent a lot of time working with Docker I've seen the community go back and forwards between small (read: Alpine) and mature (eg Debian).

I like small, but I also like well tested, high performance, and easy to understand (and therefore contribute).

This shouldn't stop any efforts to reduce the size, but we should approach with open eyes. A lot of people use this image and phusion passenger is well tested and optimised for ruby workloads. It would be a shame to lose out on this goodness for saving a miniscule amount of bandwidth when the image is updated.

So might I suggest before we make any official changes that we do some due diligence by looking also at performance.

**Comment by bethesque** [comment]
We could always keep the smaller image on a fork and release it with a separate tag to give people the option.

**Comment by k-ong** [comment]
after continued work on the alpine version of the pact broker, I do not believe that the effort is justifiable for an image in alpine. The alpine image not much smaller than it's ubuntu counterpart and could be due to the fact that we have to compile nginx and passenger from source.

the complexity and effort of maintaining the nginx and phusion passenger compilation and installation far outweighs the benefit of a few MBs of the uncompressed image.

**Comment by DerKnecht** [comment]
@k-ong how much space does the ruby:2.4-alpine take? I couldn't find it in dockerstore.
"2.4.2-alpine3.6" has 33 MB.
I really wonder what exactly makes our image so heavy. Is this just from nginx and passenger?

Though it's true that this seems to be much more effort that is needed for some MB than thought. mefellows is also right that we also should have a look at potential losses in functionality and performance when we're creating a slim-image. Even if we find a solution that needs constant maintaining that may be not better than having some hundred MB more in our image.

**Comment by YOU54F** [comment]
Hey up,

I was looking into this using a couple of tools, and noted that the expanded pact-broker image does have a-lot of wasted space, so there is room for improvement. I don't think going down the route of a pure slimmed alpine image is the right way to go, for various reasons as outlined by some above.

Will dig into this when I get more time
<table>
<thead>
<tr>
<th>Image Name</th>
<th>Compressed Size</th>
<th>Uncompressed Size</th>
<th>Potential Wasted Space</th>
<th>Image Efficiency</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://hub.docker.com/r/dius/pact-broker/">dius/pact-broker</a></td>
<td><a href="https://microbadger.com/images/dius/pact-broker">362.2 MB</a></td>
<td>1.05GB</td>
<td>452 MB</td>
<td>76 %</td>
</tr>
<tr>
<td><a href="https://hub.docker.com/r/phusion/passenger-ruby24/">phusion/passenger-ruby24</a></td>
<td><a href="https://microbadger.com/images/phusion/passenger-ruby24">239.8 MB</a></td>
<td>708MB</td>
<td>94MB</td>
<td>91%</td>
</tr>
<tr>
<td><a href="https://hub.docker.com/r/phusion/passenger-customizable/">phusion/passenger-customizable</a></td>
<td><a href="https://microbadger.com/images/phusion/passenger-customizable">199.8MB</a></td>
<td>584MB</td>
<td>45MB</td>
<td>94%</td>
</tr>
<tr>
<td>above + ruby 2.4.5 install script</td>
<td> </td>
<td>850MB</td>
<td>214 MB</td>
<td>83%</td>
</tr>
<tr>
<td><a href="https://github.com/DiUS/pact_broker-docker/blob/a40008f32f2a4b27ab347d8c16cf2ada28627db6/Dockerfile.v2">alpine slimmed</a></td>
<td>255MB</td>
<td>91MB</td>
<td>8MB</td>
<td>98%</td>
</tr>
</tbody>
</table>
<p> </p>
<p>https://microbadger.com for checking layers of compressed public images</p>
<p> </p>
<p>https://github.com/wagoodman/dive for uncompressed images, step through each layer and see the impact of each operation.</p>
<p> </p>
<p> </p>

**Comment by YOU54F** [comment]
also passenger images now with ubuntu 18.04 so the forced OS upgrade is probably unneccessary. Will get some more time soon to have a play at getting rid of some of the wasted space in the main pact-broker image as a starting point.

**Comment by bethesque** [comment]
That would be awesome, thanks.

**Comment by YOU54F** [comment]
If we cleanup on the same line as the upgrade, we get a pretty darn efficient image 👌 

```
RUN apt-get update & \
    apt-get upgrade -y -o Dpkg::Options::="--force-confold" & \
    apt-get -qy autoremove & \
    apt-get clean & \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```

```
Total Image size: 795 MB
Potential wasted space: 31 MB
Image efficiency score: 96 %
```

**Comment by YOU54F** [comment]
baseimage-docker which the main image is built on, do the same here

https://github.com/phusion/baseimage-docker/blob/master/image/bin/install_clean

**Comment by bethesque** [comment]
Awesome. Am doing a new release with the merged PR.

**Comment by bethesque** [comment]
@YOU54F, @k-ong picked up that you're actually running those commands in the background (`&`), rather than sequentially (`&&`). Here's the corrected syntax.

```
RUN apt-get update && \
    apt-get upgrade -y -o Dpkg::Options::="--force-confold" && \
    apt-get -qy autoremove && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```

Did you end up submitting that PR to the original source? You might want to update it.

**Comment by YOU54F** [comment]
Good spot @k-ong, my apologies all

I did submit a PR and will update.

& will indeed run in parallel and && will only run the subsequent command if the former passes

**Comment by bethesque** [comment]
There is a new release of the Docker Pact Broker, now at https://hub.docker.com/r/pactfoundation/pact-broker/tags *98MB compressed*! Try out the `edge` tag. When it's been battle tested for a while, I'll put it out on `latest`. It uses puma and alpine linux instead of passenger phusion.
